### PR TITLE
Allow for datetimes everywhere we allow fuzzy datetimes

### LINKED ISSUE #218: Inconsistent handling of date/times
When describing the beginning and end of a term or a bill action date, the user must use an iso format string

```python
            p.add_term('Commissioner', 'legislature', 
                       district='District {}'.format(district), 
                       start_date=start_date.isoformat(),
                       end_date=end_date)
```

When adding an event, the use needs to provide a datetime:

```
                e = Event(name=event["Name"]["label"],
                          start_time=when,
                          description=description,
                          timezone='US/Central',
                          location_name=location,
                          status=status)
```

Personally, I find this confusing. Is there any reason why we shouldn't just use a datetime.date object instead of an iso format string?

@jamesturk, @jpmckinney thoughts?

### PR DESCRIPTION
**Pull request by fgregg**
This will close #218

### CODE REVIEW

**Comment by jpmckinney** [comment]
Hmm, so:

```
fuzzy_date = {"type" : [fuzzy_date_string, "datetime"]}
```

expands to:

```
fuzzy_date = {"type" : [{"type": "string", "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}$"}, "datetime"]}
```

I don't think that works.

**Comment by fgregg** [comment]
@jpmckinney I'm not sure I understand what you mean "it doesn't work". It what sense is it not working?

File: pupa/scrape/schemas/common.py [hunk 1]
```diff
@@ -23,10 +23,20 @@
...
+fuzzy_date_string = {"type": "string", 
+                     "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}$"}
+fuzzy_date_string_blank = {"type": "string", 
+                           "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}$", 
+                           "blank": True}
+fuzzy_datetime_string_blank = {"type": "string",
+                               "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}( [0-9]{2}:[0-9]{2}:[0-9]{2})?$",
+                               "blank": True}
+
+fuzzy_date = {"type" : [fuzzy_date_string, "datetime"]}
```
Comment (fgregg) [inline pupa/scrape/schemas/common.py]: @jamesturk How could I have this validate only to "date" and not "datetime"?

**Comment by jpmckinney** [comment]
I'm not aware that this line (which is what the code expands to) means anything in validictory (though maybe validictory has features I'm unaware of).

```python
fuzzy_date = {
  "type" : [
    {
      "type": "string",
      "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}$"
    },
    "datetime"
  ]
}
```

I would be astonished if validictory can handle nested `type` like that.

**Comment by fgregg** [comment]
It can

http://validictory.readthedocs.org/en/latest/usage.html#specifying-custom-types

and it's working locally.

On Wed, Dec 16, 2015 at 2:18 PM, James McKinney <notifications@github.com>
wrote:

> I'm not aware that this line (which is what the code expands to) means
> anything in validictory (though maybe validictory has features I'm unaware
> of).
>
> fuzzy_date = {
>   "type" : [
>     {
>       "type": "string",
>       "pattern": "(^[0-9]{4})?(-[0-9]{2}){0,2}$"
>     },
>     "datetime"
>   ]
> }
>
> I would be astonished if validictory can handle nested type like that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencivicdata/pupa/pull/219#issuecomment-165228883>.
>



-- 
773.888.2718

**Comment by jpmckinney** [comment]
That's quite a surprising feature. Would seem more intuitive to use JSON Schema's `oneOf`.

**Pull request merged**